glommio/src/parking.rs

```diff
@@ -592,78 +604,62 @@ impl Reactor {
     ///
     /// Returns the duration until the next timer before this method was called.
-    fn process_timers(&self, wakers: &mut Vec<Waker>) -> Option<Duration> {
+    fn process_timers(&self) -> (Option<Duration>, usize) {
         let mut timers = self.timers.borrow_mut();
-        timers.process_timers(wakers)
+        timers.process_timers()
     }
 
-    fn process_shared_channels(&self, wakers: &mut Vec<Waker>) -> usize {
+    fn process_shared_channels(&self) -> usize {
         let mut channels = self.shared_channels.borrow_mut();
-        let mut processed = channels.process_shared_channels(wakers);
-        while let Some(w) = self.sys.foreign_notifiers() {
+        let mut processed = channels.process_shared_channels();
+        while let Some(waker) = self.sys.foreign_notifiers() {
             processed += 1;
-            wakers.push(w)
+            wake!(waker);
         }
         processed
     }
 
     fn rush_dispatch(&self, source: &Source) -> io::Result<()> {
-        let mut wakers = self.wakers.borrow_mut();
-        self.sys
-            .rush_dispatch(Some(source.latency_req()), &mut wakers)?;
-        // Wake up ready tasks.
-        for waker in wakers.drain(..) {
-            // Don't let a panicking waker blow everything up.
-            let _ = panic::catch_unwind(|| waker.wake());
-        }
+        self.sys.rush_dispatch(Some(source.latency_req()), &mut 0)?;
         Ok(())
     }
 
     /// polls for I/O, but does not change any timer registration.
     ///
     /// This doesn't ever sleep, and does not touch the preempt timer.
     pub(crate) fn spin_poll_io(&self) -> io::Result<bool> {
-        let mut wakers = self.wakers.borrow_mut();
+        let mut woke = 0;
         // any duration, just so we land in the latency ring
         self.sys
-            .rush_dispatch(Some(Latency::Matters(Duration::from_secs(1))), &mut wakers)?;
+            .rush_dispatch(Some(Latency::Matters(Duration::from_secs(1))), &mut woke)?;
         self.sys
-            .rush_dispatch(Some(Latency::NotImportant), &mut wakers)?;
-        self.sys.rush_dispatch(None, &mut wakers)?;
-        self.process_timers(&mut wakers);
-        self.process_shared_channels(&mut wakers);
-
-        let woke = wakers.len();
-        for waker in wakers.drain(..) {
-            // Don't let a panicking waker blow everything up.
-            let _ = panic::catch_unwind(|| waker.wake());
-        }
+            .rush_dispatch(Some(Latency::NotImportant), &mut woke)?;
+        self.sys.rush_dispatch(None, &mut woke)?;
+        woke += self.process_timers().1;
+        woke += self.process_shared_channels();
+
         Ok(woke > 0)
     }
 
-    fn process_external_events(&self, wakers: &mut Vec<Waker>) -> Option<Duration> {
-        let next_timer = self.process_timers(wakers);
-        self.process_shared_channels(wakers);
+    fn process_external_events(&self) -> Option<Duration> {
```

Comment:
This looks wrong to me, and likely the reason why tests aren't passing.
Nowhere you return a count of wakers from process_external events, and nowhere the fact that we have wakes here is fed to `wait`.

Keep in mind this is used in the decision of whether or not to sleep: if we had any wake, any wake at all, at any point, we must not sleep. If we woke, and sleep, we may never get to run the future and that will present itself as a hang in the test.

* if you have a waker *before* you called `wait`, you must not sleep
* if you have a waker *during* the call to `wait` (for example if you had to flush the cancellation queue, or because a remote shared channel notification atomically appeared) you must not sleep.